### PR TITLE
Pass go test flags to integration tests

### PR DESCRIPTION
**Pull request by pchila**
## Why is it important?

It allows to propagate go test flags to the actual go test command allowing for finer control over the test execution.
The `-test.short` is also important for allowing a quicker execution for PRs, for example while maintaining a full integration test run in other contexts.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- ~~[ ] I have added an integration test or an E2E test~~

## Author's Checklist

- [ ]

## How to test this PR locally

Try to run integration tests (both locally and remotely) with some custom go test flags. There are some examples listed in the [dev-guide](https://github.com/elastic/elastic-agent/blob/b2a0f2dc837a524763b893a134d552e77b3d17ff/docs/test-framework-dev-guide.md#passing-additional-go-test-flags), for example:

We pass a `-test.run` flag along with the names of the tests we want to run in OR
```
GOTEST_FLAGS="-test.run ^(TestStandaloneUpgrade|TestFleetManagedUpgrade)$" mage integration:test
```

## Related issues

- Relates #2955 

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

File: magefile.go [hunk 1]
```diff
@@ -1425,17 +1432,17 @@ func (Integration) TestOnRemote(ctx context.Context) error {
 		testPrefix := fmt.Sprintf("%s.%s", prefix, filepath.Base(packageName))
 		testName := fmt.Sprintf("remote-%s", testPrefix)
 		fileName := fmt.Sprintf("build/TEST-go-%s", testName)
+		extraFlags := make([]string, 0, len(goTestFlags)+3)
+		extraFlags = append(extraFlags, goTestFlags...)
+		extraFlags = append(extraFlags, "-test.shuffle", "on",
+			"-test.timeout", "0", "-test.run", "'^("+strings.Join(packageTests, "|")+")$'")
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply magefile.go:1438]: +1 should be `+6`

**Review by blakerouse — APPROVED** [review]
Change looks good to me.

File: magefile.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pchila) [reply magefile.go]: good catch! This is a leftover from when I was thinking about changing the flag form to `-test.foo=bar`... Will change to 6 :smile:

**Comment by pchila** [comment]
ok,  now integration tests are running correctly after a small change in 5c3227092f2e6b53bdd19a0d026a598558fc3fbb

/cc @ycombinator

File: dev-tools/mage/gotest.go [hunk 2]
```diff
@@ -232,11 +232,10 @@ func GoTest(ctx context.Context, params GoTestArgs) error {
 		)
 	}
 
+	testArgs = append(testArgs, params.ExtraFlags...)
```
Comment (ycombinator) [inline dev-tools/mage/gotest.go:235]: Just curious: why was this change needed?

File: docs/test-framework-dev-guide.md [hunk 3]
```diff
@@ -49,6 +49,41 @@ pass `[testName]` to `go test` as `--run=[testName]`.
...
+A few examples:
+
+##### Run a single test with an exact match
+We want to run only the test named "TestStandaloneUpgrade"
+`GOTEST_FLAGS="-test.run ^TestStandaloneUpgrade$" mage integration:test`
+
+##### Run a tests matching a partial expression
+We want to run any test with "Upgrade" in the name
+`GOTEST_FLAGS="-test.run Upgrade" mage integration:test`
+##### Run a single test and signal that we want the short version
```
Comment (ycombinator) [inline docs/test-framework-dev-guide.md:71]:
```suggestion

##### Run a single test and signal that we want the short version
```

File: docs/test-framework-dev-guide.md [hunk 4]
```diff
@@ -49,6 +49,41 @@ pass `[testName]` to `go test` as `--run=[testName]`.
...
+We pass a `-test.count` flag along with the name match
+`GOTEST_FLAGS="-test.run ^TestStandaloneUpgrade$ -test.count 10" mage integration:test`
+
+##### Run specific tests
+We pass a `-test.run` flag along with the names of the tests we want to run in OR
+`GOTEST_FLAGS="-test.run ^(TestStandaloneUpgrade|TestFleetManagedUpgrade)$" mage integration:test`
+
+##### Limitations
+Due to the way the parameters are passed to `devtools.GoTest` the value of the environment variable
+is split on space, so not all combination of flags and their values may be correctly split.
```
Comment (ycombinator) [inline docs/test-framework-dev-guide.md:85]: Thanks for adding this section.

Does this limitation have something to do with some go test flags taking arguments, e.g. `-test.run <test name regex>`, vs. others not taking arguments, e.g. `-short`?

File: magefile.go [hunk 5]
```diff
@@ -1491,7 +1506,7 @@ func integRunner(ctx context.Context, matrix bool, singleTest string) error {
 	return nil
 }
 
-func createTestRunner(matrix bool, singleTest string, batches ...define.Batch) (*runner.Runner, error) {
+func createTestRunner(matrix bool, singleTest string, goTestFlags string, batches ...define.Batch) (*runner.Runner, error) {
```
[the remark this answers is not in the document]
Comment (pchila) [reply magefile.go:1509]: We could... I am leaving it in, in case the new flags do not behave like we want but we can deprecate them and create a separate issue to clean up

File: docs/test-framework-dev-guide.md (hunk 4 above)
Comment (pchila) [reply docs/test-framework-dev-guide.md:85]: It does not have any problems with flags that take arguments or not... 
The problem is that when one argument contains a space, it will be split incorrectly.
The correct way would be to:
- either reimplement the flag parser of `go test` (those are not exposed sadly and reimplementing it and keeping it in sync with the future versions of go it will be probably too much work)
- or define a completely different set of flags and translate them into go test flags (kinda misses the point of the PR)

Another alternative is to build the `go test` command differently so that we can plonk the flags as a single string being completely agnostic about the contents but that is also a bit of work...

For this PR I went with the simple and stupid solution and I couldn't think of a scenario where I needed a smarter parsing of flags: we'll cross that bridge when we have to.

File: magefile.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (blakerouse) [reply magefile.go:1509]: I like `mage integration:single [testName]` it is much clean CLI command to run a test than all the ENV flags being set. I wish we could get mage to take command line arguments for targets. I would prefer to see something like:

`mage integration:test -short -name SingleTestName -platform linux/amd64`

File: dev-tools/mage/gotest.go (hunk 2 above)
Comment (pchila) [reply dev-tools/mage/gotest.go:235]: Sorry for the incoming wall of text :sweat_smile: 

TL;DR: This was needed to make sure that a `-test.run` flag specified in the `GOTEST_FLAGS` environment variable does not interfere with the batching done by the framework.

Full explanation:
The integration test framework runs the tests twice:
- the first time we pass a special tag that make all the `define` statements in the tests skip the test and dump the requirements. This output is processed by the integration test framework to discover the tests and the set of environments/machines we will need to spawn and allocate the tests to the various machines.
- the second time we run the tests (here) the integration test framework adds a `-test.run` flag when launching go test on the remote machine to make sure that only the tests corresponding to that batch are executed.

By specifying the extra flags before the `-test.run` for the batch we make sure that the last flag definition "wins" (have a look at the unit test in `batch_test.go`) so that whatever `run` constraint is specified in GOTEST_FLAGS participated in the discovery and batching (1st `go test` execution) but doesn't override the actual execution on the remote machine.

File: magefile.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ycombinator) [reply magefile.go:1509]: Okay, then could we eliminate _just_ the `singleTest` argument to this function and have `mage integration:single [testName]` internally setup `goTestFlags` appropriately?

File: docs/test-framework-dev-guide.md (hunk 4 above)
Comment (ycombinator) [reply docs/test-framework-dev-guide.md:86]: Understood, thanks for clarifying.  I'm good with the simple solution for now.

File: dev-tools/mage/gotest.go (hunk 2 above)
Comment (cmacknz) [reply dev-tools/mage/gotest.go:235]: This should definitely be a comment in the code.

Comment (pchila) [reply dev-tools/mage/gotest.go:251]: Added the comment just above the change.

**Comment by pchila** [comment]
buildkite test this
